File: 108-convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.py
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def helper(l,r):
            if(l>r):
                return None
            m=(l+r)//2
            root=TreeNode(nums[m])
            root.left=helper(l,m-1)
            root.right=helper(m+1,r)
            return root
        return helper(0,len(nums)-1)
# # T= O(n log n),S=O(n)
# class Solution:
#     def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
#         total_nums = len(nums)
#         if not total_nums:
#             return None

#         mid_node = total_nums // 2
#         return TreeNode(
#             nums[mid_node], 
#             self.sortedArrayToBST(nums[:mid_node]), self.sortedArrayToBST(nums[mid_node + 1 :])
#         )
# class Solution:
#     def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
#         return self.makeBST(nums, 0, len(nums))
    
#     def makeBST(self, nums, start, end):
#         if start >= end: return None
#         return TreeNode(
#             val=nums[ (start + end)//2 ],
#             left=self.makeBST(nums, start, (start + end)//2),
#             right=self.makeBST(nums, 1+((start+end)//2), end)
#         )

File: 108-convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.java
class Solution {
    public TreeNode helper(int[]nums,int lo,int hi){
        if(lo>hi) return null;
        int mid=(lo+hi)/2;
        TreeNode root=new TreeNode(nums[mid]);
        root.left=helper(nums,lo,mid-1);
        root.right=helper(nums,mid+1,hi);
        return root;
    }
    public TreeNode sortedArrayToBST(int[] nums) {
        int n=nums.length;
        return helper(nums,0,n-1);
    }
}